programs/bpf_loader/src/syscalls.rs

```diff
@@ -2204,7 +2205,7 @@ fn call<'a>(
             *account_ref.lamports = account.lamports();
             *account_ref.owner = *account.owner();
             if account_ref.data.len() != account.data().len() {
-                if !account_ref.data.is_empty() {
+                if !do_support_realloc && !account_ref.data.is_empty() {
                     // Only support for `CreateAccount` at this time.
```

Comment:
test case here is below where `X > 0 && X < MAX_PERMITTED_DATA_LENGTH`

1. program A accepts two self-owned account inputs A and B (`data.len() == X` but zeroed)
2. program A reallocs account input A to a size of `data.len() + X`
3. program A self-invokes with account input A
    a. program A then reallocs that account to a total size of `data.len() + X + MAX_PERMITTED_DATA_INCREASE`
    b. program A then sets all data to non-zero values
    c. program A returns to depth 0
4. program A reallocs account input A to total size `data.len() + MAX_PERMITTED_DATA_INCREASE`

test case should assert that account B is still zeroed

Reply:
The only suspicious part of that scenario as far as I can tell is the realloc greater than the max size but that will be caught during deserialization.  Mishandling the realloc and overwriting serialized data is a footgun for sure but not a security risk and that footgun already exists before these changes.

Reply:
This case just highlights that if a program reallocs an account itself and then does a CPI which also reallocs, the runtime will overwrite serialized data. I feel that this would be unexpected. And it could go undetected if the program then reallocs the account again to be within the permitted buffer by the time it returns

Reply:
The problem is that the account data size at the time of the cpi call is used for the max increase check instead of the original size at the time of serialization.  This could lead to an undetected overalloc.  I've changed the check to use the original data size and added a test.  Thanks @jstarry !